cloudpathlib/cloudpath.py

```diff
@@ -773,9 +792,16 @@ def copytree(
                 "Destination path {destination} of copytree must be a directory."
             )
 
+        if ignore is not None:
+            ignored_names = ignore(self.fspath, [x.name for x in self.iterdir()])
```

Comment:
Two thoughts here:

(1) I think that we may want `self.c._no_prefix_no_drive` instead, which removes the `s3://bucket` portion but keeps the rest. This way the first argument won't include the path to the potentially arbitrary local folder in the cache.

(2) I know it follows the CPython implementation, but it's not ideal to call `iterdir` twice since there's network call overhead on both of them. Would it be feasible to do something like the following within the main loop instead?

```python
...

for subpath in self.iterdir():
    if not ignore(subpath.parent._no_drive, [subpath.name]):
        continue
    ...  
```

Maybe not ideal if calls to `ignore` are expensive, but there's a tradeoff with the network overhead of `iterdir` that may be worth it.

Reply:
Just looked a little closer at the CPython one—at least point 2 we could address like their implementation where we introduce a `_copytree` that gets passed `entries` so that we don't do the expensive operation twice (like they do to avoid calling `scandir` twice)

Reply:
Yeah, I get it. Is not a bad idea since it replaces a single call on every file for multiple calls on a single file each time, but keeping the interface untouched. Would it be a bad idea to cache or store the result of `self.iterdir()` and use it twice? 

```python
contents = list(self.iterdir())  # Using list because iterdir is a generator

if ignore is not None:
    ignored_names = ignore(subpath.parent._no_prefix_no_drive, [x.name for x in contents])
else:
    ignored_names = set()

# ...

for subpath in contents:
    # ...
```

Reply:
Seems like what I proposed is equivalent to what you described. They do the step of creating a list of entries [here](https://github.com/python/cpython/blob/a4ac14faa5c2be433738dfbbed14b0eaa1a2399e/Lib/shutil.py#L560)

Reply:
👍 Oh yeah, that looks simple and saves us some network overhead! Let's do it

Reply:
Ah, maybe we should still use their solution. `copytree` is a recursive function, so we still are making network calls (at least is half now). Seems like the only way to make a single network call is using their approach to determine the list of entries before actually calling `copytree` (which will be named `_copytree` then)

Reply:
Also, the recursive call (when `subpath` is a directory) also needs the `ignore` argument passed

Reply:
> Ah, maybe we should still use their solution. `copytree` is a recursive function, so we still are making network calls (at least is half now). Seems like the only way to make a single network call is using their approach to determine the list of entries before actually calling `copytree` (which will be named `_copytree` then)

NVM, seems like this won't make a difference since `iterdir` is not recursive, so we actually need to call `iterdir` on each depth level of the recursive function

Reply:
> Also, the recursive call (when `subpath` is a directory) also needs the `ignore` argument passed

Fixed in the latest push